strawberry/schema/execute.py

```diff
@@ -157,28 +162,36 @@ def execute_sync(
             if execution_context.errors:
                 return ExecutionResult(data=None, errors=execution_context.errors)
 
-        result = original_execute(
-            schema,
-            execution_context.graphql_document,
-            root_value=execution_context.root_value,
-            middleware=extensions_runner.as_middleware_manager(),
-            variable_values=execution_context.variables,
-            operation_name=execution_context.operation_name,
-            context_value=execution_context.context,
-            execution_context_class=execution_context_class,
-        )
-
-        if isawaitable(result):
-            ensure_future(cast(Awaitable[GraphQLExecutionResult], result)).cancel()
-            raise RuntimeError("GraphQL execution failed to complete synchronously.")
+        with extensions_runner.executing():
+            if not execution_context.result:
+                result = original_execute(
+                    schema,
+                    execution_context.graphql_document,
+                    root_value=execution_context.root_value,
+                    middleware=extensions_runner.as_middleware_manager(),
+                    variable_values=execution_context.variables,
+                    operation_name=execution_context.operation_name,
+                    context_value=execution_context.context,
+                    execution_context_class=execution_context_class,
+                )
 
-        result = cast(GraphQLExecutionResult, result)
-        execution_context.result = result
-        if result.errors:
-            execution_context.errors = result.errors
+                if isawaitable(result):
+                    ensure_future(
+                        cast(Awaitable[GraphQLExecutionResult], result)
+                    ).cancel()
```

Comment:
```suggestion
                if isawaitable(result):
                    result = cast(Awaitable[GraphQLExecutionResult], result)
                    ensure_future(result).cancel()
```

Reply:
While we're at it we could also change `ensure_future` to `create_task`, which appears to be the [preferred variant in 3.7](https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task) :)